text/2023-04-26-color-outlines.md

```diff
@@ -0,0 +1,198 @@
+# API for color outlines
+
+In order to support [COLRv1] (and v0) fonts, the
+[skrifa](https://github.com/googlefonts/fontations/tree/main/skrifa)
+crate requires an API to expose a more sophisticated set of operations than
+are handled by the current [`Pen`](https://docs.rs/font-types/0.1.4/font_types/trait.Pen.html)
+trait that is used to generate basic outlines.
+
+## Proposal
+
+The color [specification][COLRv1] defines a directed acyclic graph of "paint"
+nodes with each describing an operation such as transform, clip or fill. An
+implementation is expected to traverse the graph in depth first order applying
+those operations in a sequence. While this is a reasonable representation for
+storage and analysis, exposing this data structure directly through a trait
+based callback interface is cumbersome.
+
+The proposed API would flatten the graph structure into a linear sequence of
+drawing commands, making use of stack semantics to capture the graph structure
+where necessary. The set of commands required to represent the graph is fairly
+small and maps nicely to most graphics APIs which already represent layering
+with a state stack.
+
+## Interface
+
+This is a rough outline of the proposed interface which would introduce a new
+`ColorPen` trait for accessing color outlines.
+
+### Supporting types
+
+Basic types that are used by the new `ColorPen` trait.
+
+```rust
+/// Extents of a rectangular region.
+pub struct BoundingBox;
+
+/// Affine transformation matrix.
+pub struct Transform;
+
+/// 32-bit RGBA unpremultiplied color.
+pub struct Color;
+
+/// Enumeration containing data for solid or gradient fills.
+pub enum Brush<'a>;
```

Comment:
As discussed, it would be useful to see a bit more detail here on the value type and coordinates used here. Making this work well and be easily adaptable with the current Skia implementation would mean that ideally they'd stay quite close to what's in the font. 

I think the adaptations to Cairo or Direct2D patterns/shaders and their constraints are different to the ones for Skia, for example.

Reply:
I’ll add some definitions for this type. In the current prototype, the structure and values should exactly match those returned by FreeType with two exceptions: we do the 16.16 to f32 conversion for all values and the line projection for linear gradients (using the same code as Skia but without the y flip).